python/cudf/udf_cpp/nrt/_nrt_cuda.cuh

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright (c) 2012, Anaconda, Inc.
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are
+met:
+
+Redistributions of source code must retain the above copyright notice,
+this list of conditions and the following disclaimer.
+
+Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <cuda/atomic>
+
+typedef __device__ void (*NRT_dtor_function)(void* ptr, size_t size, void* info);
+typedef __device__ void (*NRT_dealloc_func)(void* ptr, void* dealloc_info);
+
+typedef __device__ void* (*NRT_malloc_func)(size_t size);
+typedef __device__ void* (*NRT_realloc_func)(void* ptr, size_t new_size);
+typedef __device__ void (*NRT_free_func)(void* ptr);
+
+extern "C" {
+struct MemInfo {
+  cuda::atomic<size_t, cuda::thread_scope_device> refct;
+  NRT_dtor_function dtor;
+  void* dtor_info;
+  void* data;
+  size_t size;
+};
+}
+
+typedef struct MemInfo NRT_MemInfo;
+
+// Globally needed variables
+struct NRT_MemSys {
+  /* System allocation functions */
+  struct {
+    NRT_malloc_func malloc;
+    NRT_realloc_func realloc;
+    NRT_free_func free;
+  } allocator;
+  struct {
+    bool enabled;
+    cuda::atomic<size_t, cuda::thread_scope_device> alloc;
+    cuda::atomic<size_t, cuda::thread_scope_device> free;
+    cuda::atomic<size_t, cuda::thread_scope_device> mi_alloc;
+    cuda::atomic<size_t, cuda::thread_scope_device> mi_free;
+  } stats;
+};
+
+__device__ void* malloc_wrapper(size_t size) { return malloc(size); }
+
+__device__ void free_wrapper(void* ptr) { free(ptr); }
+
+/* The Memory System object */
+__device__ NRT_MemSys TheMSys = {
```

Comment:
If this header is included in two separate translation units, will `TheMSys` be multiply-defined?

Reply:
I added an include guard for this scenario in https://github.com/rapidsai/cudf/pull/13060/commits/d9001d3a390e081c9b20b35d67de63e4e523d4e7

Reply:
I don't think an include guard will stop the symbol being defined twice in separate translation units - it can be included once in each TU.

Reply:
ah I see what you mean. Working on a fix for this.